inference-engine/src/preprocessing/ie_preprocess_gapi_kernels.cpp

```diff
@@ -42,572 +42,529 @@
+        return [](const std::array<const uint8_t*, chs>& ins, uint8_t* out, const int length) {
+            const auto inT = reinterpret_cast<const std::array<const float*, chs>&>(ins);
+            auto outT = reinterpret_cast<float*>(out);
+            isa_tag_t t;
+            mergeRowImpl(t, inT, outT, length);
+        };
+    }
+#endif
+};
 
-        if (std::is_same<T, uint8_t>::value && chs == 3) {
-            splitRow_8UC3(in, outs[0], outs[1], outs[2], length);
-            return;
-        }
+}  // namespace
 
-        if (std::is_same<T, uint8_t>::value && chs == 4) {
-            splitRow_8UC4(in, outs[0], outs[1], outs[2], outs[3], length);
-            return;
-        }
+namespace {
+using split_supported_types = typelist<uint8_t, int8_t, uint16_t, int16_t, int32_t, float, fp_16_t>;
 
-        if (std::is_same<T, float>::value && chs == 2) {
-            splitRow_32FC2(reinterpret_cast<const float*>(in),
-                           reinterpret_cast<float*>(outs[0]),
-                           reinterpret_cast<float*>(outs[1]),
-                           length);
-            return;
+template<typename T, int chs>
+void splitRowImpl(scalar_tag, const T* in, std::array<T*, chs>& outs, const int length) {
+    for (int x = 0; x < length; ++x) {
+        for (int c = 0; c < chs; ++c) {
+            outs[c][x] = in[chs * x + c];
         }
+    }
+}
 
-        if (std::is_same<T, float>::value && chs == 3) {
-            splitRow_32FC3(reinterpret_cast<const float*>(in),
-                           reinterpret_cast<float*>(outs[0]),
-                           reinterpret_cast<float*>(outs[1]),
-                           reinterpret_cast<float*>(outs[2]),
-                           length);
-            return;
-        }
+template<typename isa_tag_t, int chs>
+struct typed_split_row {
+    using p_f = void (*)(const uint8_t* in, std::array<uint8_t*, chs>& outs, const int length);
 
-        if (std::is_same<T, float>::value && chs == 4) {
-            splitRow_32FC4(reinterpret_cast<const float*>(in),
-                           reinterpret_cast<float*>(outs[0]),
-                           reinterpret_cast<float*>(outs[1]),
-                           reinterpret_cast<float*>(outs[2]),
-                           reinterpret_cast<float*>(outs[3]),
-                           length);
-            return;
-        }
+    template <typename type>
+    p_f operator()(type_to_type<type> ) {
+        return [](const uint8_t* in, std::array<uint8_t*, chs>& outs, const int length) {
+            const auto inT = reinterpret_cast<const type*>(in);
+            auto outT = reinterpret_cast<std::array<type*, chs>&>(outs);
+            scalar_tag t;
+            splitRowImpl(t, inT, outT, length);
+        };
     }
-#endif  // HAVE_SSE
 
-#ifdef HAVE_NEON
-    if (std::is_same<T, uint8_t>::value && chs == 2) {
-        neon::splitRow_8UC2(in, outs[0], outs[1], length);
-        return;
+#if MANUAL_SIMD
```

Comment:
@AsyaPronina @rgarnov I think that there is possibility to replace conditional compilation macro the std::enable_if. However I don't have idea how do it better for this case exactly. Could you please give me an advice?